Update staging environment

matthiasrohmer opened the pull request:
Still rough around the edges. This introduces workflows to create two type of staging environments, using a gulp task which is called by a GitHub actions job step to orchestrate all.

After a PR is synced (created, pushed to or deleted) the check workflow starts. This PR adds a new job to that workflow: `stage`. This is called in parallel to the lint and test jobs and does four things:

1. Post a sticky comment to the PR the event came from that a staging build is in progress. Sticky means an eventually existing comments gets replaced.
2. Determine the changed files in the PR, which are split into two categories: files that relate to the static build output (`static`) and files that have impact on the serving App Engine instance (`app`).
3. Start a gulp task that calls a Google Cloud Build webhook which, depending on the changed files, either starts a job for a full deploy (`app`, `gcloud app deploy`) or one to build the static files to a Google Cloud Storage bucket.
3.1. After the Cloud Build job is started, the gulp task repeatedly (fixed cliff, then every 60s) queries the GitHub checks API for the result of the Google Cloud Build check.
3.2. As soon as it has completed, the check details together with the changed files are used to create a helpful comment including a link to the deployed preview and deeplinks to the changed pages.
4. Update the sticky comment with the one created by the gulp task.

While this is working there are some things to be improved:

- [x] For the static builds to be working, https://github.com/GoogleChrome/developer.chrome.com/pull/4383 needs to be merged
- [x] Timings are not yet great. It might make sense to call the webhook outside of the gulp task, that would skip the Node.js install steps. A potential 40s win.
- [x] Look into if the webook's own performance can be improved as it adds another 1m30s upfront
- [ ] Better error handling: if the job early exits for whatever reason there might be a dangling "Staging build in progress ..." comment on PRs
- [ ] Clean up: after PR-close built pages and instances need to be removed, after a push to a PR the previous preview needs to be removed. Question: instead of commit SHAs use PR numbers for versioning? Then we would simply overwrite, but loose comparability

devnook reviewed (approved):
This looks all reasonable, I left mostly comments to just better understand the details. Wa can also discuss them in person if thats easier.

Should #4383 get merged before we continue with this one? If yes, can you undraft it?

Regarding previous version cleanup, I think per-PR granularity is acceptable, because this is how Netlify works right now. Sometimes it is good to see which commit caused which change, though. We can go either way I think.

matthiasrohmer commented on .cloudbuild/stage-static.yaml:
```diff
@@ -0,0 +1,23 @@
+      - |
+        cp -rf dist/en/* dist/ &&
+        rm -rf dist/en &&
+        gsutil -q -m cp -r dist gs://static-dcc-staging/$SHORT_SHA &&
```
Yes, gsutil proved to be around 30s faster for the same amount of files 🙂  the new `cloud storage` commands seem to focus on performance for bigger files.

matthiasrohmer commented on .cloudbuild/stage.yaml:
```diff
@@ -1,24 +1,13 @@
-        cp -rf dist/en/* dist/ &&
-        rm -rf dist/en &&
-        gcloud storage cp -L manifest.csv -j html,xml,css,js --recursive dist gs://static-dcc-staging/$SHORT_SHA &&
-        gcloud storage cp manifest.csv gs://static-dcc-staging/$SHORT_SHA/manifest.csv &&
```
I had the idea of using the manifest.csv as a lookup table for an `fs.access` alternative. All the redirect handlers we have are looking up files on the file system and I thought this approach might save us some Cloud Storage API calls and by this latency. But `gsutil` doesn't have the manifest feature.

matthiasrohmer commented on .github/workflows/check.yml:
```diff
@@ -91,6 +91,57 @@ jobs:
+          filters: |
+            app:
+              - 'package.json'
+              - 'server/**/*.js'
```
Good point! I had a feeling filtering for server might be too little. Let me know if there is another watch target that comes to your mind 🙂

matthiasrohmer commented on .github/workflows/check.yml:
```diff
@@ -91,6 +91,57 @@ jobs:
+      - name: Staging
+        id: staging
+        env:
+          CHANGED_APP_FILES: ${{ steps.changes.outputs.app_files }}
```
Yep indeed!

> If enabled, for each filter it sets an output variable with the name ${FILTER_NAME}_files. It will contain a list of all files matching the filter

from https://github.com/dorny/paths-filter#outputs.

matthiasrohmer replied:
Not using dorny/paths-filter anymore as this now requires a file-changed method that works both on Cloud Build and GitHub actions, so I am simply querying the GitHub API and using gulp tasks to filter.

matthiasrohmer commented on gulp-tasks/stagePr.js:
```diff
@@ -0,0 +1,281 @@
+  // Passing output from logs is deprecated and streaming to the
+  // output environment variable would require logs to be less verbose
+  // https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
+  await fs.writeFile('announcement.md', announcement, {encoding: 'utf-8'});
```
Will do! I'll probably also rephrase the comment to make it clear that file is temporary.

matthiasrohmer replied:
All temp files are now written to gulp-tasks/stageGitHub/tmp, which is also in the .gitignore.

matthiasrohmer commented on gulp-tasks/stagePr.js:
```diff
@@ -0,0 +1,281 @@
+    `**${baseUrl}**`;
+
+  const changedPages = [];
+  if (changedFiles.static) {
```
Yeah it got complex... I will see what is testable without messing with the external deps too much.

matthiasrohmer commented on gulp-tasks/stagePr.js:
```diff
@@ -0,0 +1,281 @@
+async function stagePr() {
+  if (!process.env.GITHUB_ACTION || !process.env.COMMIT_SHA) {
+    console.warn(
+      'This task is inteded to run on GitHub actions. Use npm run stage:personal locally instead.'
```
Good point, will do!

matthiasrohmer replied:
I split up the task into multiple smaller ones and also scoped them to their own gulpfile, so no one running `gulp` would accidentally stumble over them.

matthiasrohmer commented on gulp-tasks/stagePr.js:
```diff
@@ -0,0 +1,281 @@
+  console.log(`Requesting staging build for commit ${commit}`);
+
+  const request = await fetch(
+    `https://cloudbuild.googleapis.com/v1/projects/dcc-staging/triggers/Webhook:webhook?key=${process.env.CLOUD_BUILD_KEY}&secret=${process.env.CLOUD_BUILD_SECRET}`,
```
I think so. At least I couldn't find any documentation on it being prohibited and the console did let me do it. Worth asking on YAQS for example?

matthiasrohmer replied:
Not using the Webhook anymore.

matthiasrohmer commented on gulp-tasks/stagePr.js:
```diff
@@ -0,0 +1,281 @@
+  // This can take up to 1m30s - wait for 2m to have room to wiggle
+  // and as the build takes a while anyway
+  console.log('Waiting for Cloud Build job to start ...');
+  await wait(2 * 60 * 1000);
```
Puh... yeah. I wanted to do this initially but didn't work towards this as the Cloud Build machines are actually beefier than the GitHub ones. That benefit gets lost with the webhook delay though.

If we can get rid of the webhook delay it might be possible the other way round: then we could think about using the Google Cloud build on GitHub as well. Not too sure about transfer times though and it would involve authenticating GitHub to the storage bucket.

matthiasrohmer commented on .github/workflows/check.yml:
```diff
@@ -91,6 +91,57 @@ jobs:
+          list-files: json
+          filters: |
+            app:
+              - 'package.json'
```
dorny/paths-filter always matches the PRs current state. So if a file has been touched in a previous commit, but reverted it wouldn't trigger the filter.

matthiasrohmer commented on .github/workflows/check.yml:
```diff
@@ -91,6 +91,57 @@ jobs:
         run: |
           npm run test
 
+  stage:
```
PR deletion (or close rather) do not trigger the check workflow, currently PR creation and subsequent pushes to the PR's branch do.

matthiasrohmer commented:
This is now ready for a final review. Fixed some late flaws and also got rid of the webhook, which makes it easier to follow the flow, on the other hand required to move some logic to Google Cloud Build.

The gulptask is now also split into three separate tasks, for easier maintenance.

See test PRs at https://github.com/GoogleChrome/developer.chrome.com/pull/4687 (static build) and https://github.com/GoogleChrome/developer.chrome.com/pull/4688 (full instance build).

This also updates the new bucket static handler to not call `next()` in case of error as this would execute the old not-found handler, which expects `dist/404.html` to exist, but the instance doesn't have the static files anymore. As discussed this also moves away from commit hashes as a versioning tool but uses `main` for the evergreen version (which Sanity can also build to) and `pr-<number>` for versions built by a PR.

The only thing that's missing is a cleanup task for the bucket and the instance versions, but that's something for the new year.

devnook commented on server/handlers/static/bucket.js:
```diff
@@ -94,21 +66,18 @@ async function bucketHandler(req, res, next) {
   } catch (e) {
     console.error('Could not serve', filePath);
-    return next();
+    res.send(`Could not send ${filePath} due to ${e}.`);
```
This is probably fine for dev / stage, but if we're about to use this handler in prod at some point we should just return 404/50X

matthiasrohmer replied:
Yep! Will create a bunch of follow-up issues once this is in, like response times can also be improved in general, redirect handlers need to work with Cloud Storage, ...

devnook commented on gulp-tasks/stageGitHub/lib/fetchGitHubApi.js:
```diff
@@ -0,0 +1,43 @@
+
+  const request = await fetch(url, {
+    headers: {
+      Authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
```
Might be worth checking if this variable exists early on and warn / return ealy if not, rather than timeout on wrong request. I'm sure someone will try to run it outside of github at some point.

matthiasrohmer replied:
I see... my world has rainbows and unicorns though. Nevertheless now throws if not authenticated 🙂

devnook commented on gulp-tasks/stageGitHub/determineDeploymentType.js:
```diff
@@ -0,0 +1,77 @@
+    return;
+  }
+
+  let changedFiles = await fetchGitHubApi(
```
Since this runs on cloud, does it know github_token variable?

devnook replied:
ah, i see it is passed in the workflow file now

matthiasrohmer replied:
Yes, I added a token to secret manager. See changes in .cloudbuild/stage.yaml.

devnook commented on gulp-tasks/stageGitHub/determineDeploymentType.js:
```diff
@@ -0,0 +1,77 @@
+  // Google Cloud Build, except the file system as of 12/2022. See:
+  // https://cloud.google.com/build/docs/configuring-builds/pass-data-between-steps
+  await fs.writeFile(DEPLOYMENT_TYPE_PATH, output, {encoding: 'utf-8'});
+  console.log(output);
```
debugging?

matthiasrohmer replied:
More or less. I wanted to signal the task actually did something, but adopted the same pattern as in the other tasks of explicitly stating what happened.

nie357 reviewed (commented):
1. ###  

- [ ] `@_****_`
